### DISCUSSION TITLE
Add notification/alert as part of the keymap.

### DISCUSSION THREAD

**Discussion by vincevargadev** [helix-editor/helix#6480] [Q&A] [answered]
I'm new to Helix and I really enjoy working with it so far. I prefer its keymappings over Vim's but unfortunately, muscle memory kicks in, and sometimes it's hard to remember how to achieve something in Helix.

For this reason, I remapped the `$` sign to jump to the end of the line (and I have many similar remaps in my config).

```toml
# The default way of going to the end of the line with Helix is: gl
"$" = "goto_line_end"
```

This is convenient because when I forget the "idiomatic Helix" binding, I can quickly achieve what I want, but I'd like to slowly train myself to switch to the Helix approach.

What I'd love to do with my new, advanced mapping is:
1. perform the original action
2. AND display some kind of warning that remind me of the "correct" way to do this action
  * this warning could be visual, e.g. a notification in the top right corner, but any other location would be acceptable (so a shell based hack is better than nothing)
  * ideally, there should also be a "bell" sound so that I remember to read the notification

In pseudo code:

```toml
"$" = ["goto_line_end", "display-toast-message: next time use gl to go to the end of the line", "bell-sound"]
```

If you have trouble understanding what I want: I'd like to write myself a "Helix Key Promoter" helper (similarly to [IntelliJ's Key Promoter](https://plugins.jetbrains.com/plugin/9792-key-promoter-x)) that helps me learn the Helix way of doing things.

**Comment by vincevargadev** [comment]
I think I got something that's reasonably good:

```toml
0 = ["goto_line_start", ":run-shell-command echo 'helix-hint(gh): use gh to go to the start of the line'"]
```

I'll keep refining this to make it more "announcement-like" and easier to read.

![Screenshot 2023-03-29 at 20 12 57](https://user-images.githubusercontent.com/9061239/228630142-2e3dd77f-8e16-4e01-a0a6-fa2c336c65d7.png)

It's a bit inconvenient that I need to hit `esc` to get rid of the hint.

**Comment by vincevargadev** [comment, marked as answer]
You can use the ASCII box drawing characters to build the outline, as well as emojis.

I wrote a simple Dart program (similar possible with shell, if you don't have Dart installed) that takes the first word as command, and the rest as explanation and displays it in a box within Helix:

```dart
void main(List<String> args) {
    final cm = args[0];
    final rest = args.sublist(1).join(' ');

    print('┌────┬─' + '─' * cm.length + '─┬─' + '─' * rest.length + '─┐');
    print('│ 🧬 │ $cm │ $rest │');
    print('└────┴─' + '─' * cm.length + '─┴─' + '─' * rest.length + '─┘');
}
```

With the following config:

```toml
# The default way of going to the start of the line with Helix is: gh
0 = ["goto_line_start", ":sh dart ~/helix_trainer.dart gh go to line start"]
# The default way of going to the end of the line with Helix is: gl
"$" = ["goto_line_end", ":sh dart ~/helix_trainer.dart gl go to line end"]
# The default way of going to the first non-blank char in line with Helix is: gs
"^" = ["goto_first_nonwhitespace", ":sh dart ~/helix_trainer.dart gs go to first non-blank in line"]
```

This is how it looks like:

https://user-images.githubusercontent.com/9061239/228643569-22d655e2-d314-49ea-b3ed-2d8fc8bd7f0a.mov

**Reply by sweetbbak** [reply]
I see you answered your own question but I just wanted to say that this is freaking cool. I'm gonna have to check out that gist and jack that file picker as well lol
